### ISSUE TITLE
Proposal: public API for hidden-state extraction (residual stream) + tools

### ISSUE THREAD

**Issue by someaka** [ggml-org/llama.cpp#27112]
## Motivation

Interpretability and analysis work on LLMs needs access to per-layer residual-stream states during normal decode. Upstream exposes embeddings (`llama_set_embeddings`) and logits, but there is no public way to read intermediate post-block hidden states. Projects that need them today carry private forks.

**Not redundant with the internal `llama_set_embeddings_layer_inp`:** that API (src/llama-ext.h, 12 model files) captures the pre-norm layer *input* (`inpL`) for spec-decode acceptance checks. What we need — and propose — is the post-block *output* residual, exposed through the public `include/llama.h` API.

## Proposed API

```c
void    llama_set_extract_hidden_states(struct llama_context * ctx, bool value);
size_t  llama_get_hidden_state_n(const struct llama_context * ctx);            // states captured
float * llama_get_hidden_state(struct llama_context * ctx, size_t i);          // last-token state i
// + per-token variants (ith / n_tokens) and a batch getter for all tokens × layers
```

Plus, if there's interest:
- `tools/hs-extract` — single-prompt JSON extraction CLI
- `tools/hs-extract-batch` — high-throughput batch extraction (streaming accumulator, checkpoint/resume)
- `POST /hidden-states` server endpoint (forward pass → per-layer vectors, no generation)

## Use case

Emotion-vector extraction over 200K+ prompts across multiple architectures (our research pipeline, cross-validated CUDA/Vulkan) — currently maintained as a fork. Happy to split into separate focused PRs (core API first, tools/endpoint later) per maintainer preference.

Gauging interest before proceeding with the PRs.

**Comment by Extremity** [comment]
I have zero to contribute here but since you're gauging interest, this is interest.

Edit: I grabbed the commit from your PR and have been using it and have not had a single significant problem. I had to make some minor edits (a few conditionals to swap things around when running in windows, a few small changes/additions for layer 64 for training, but that's it). It's been fully functional otherwise and the hs-extract-batch binary has been a great help. It's leagues ahead of the alternatives in terms of ease of use.

I'm kind of shocked this didn't receive more attention, it's super useful and - at least the parts I've used - are pretty much fully functional. I'm going to have to do a bit more modification to change the default "batch = context size" thing to make it a bit more forgiving on consumer hardware, and I wish the --batch-size flag was functional and will try to implement that, but I just wanted to say thanks - the specific stuff I'm doing required me to grab hidden states and I was having a hell of a time getting speculators/vllm/etc to do what I wanted on my setup - this changed everything.

so, again - this is somebody expressing interest. I'll be keeping a llama.cpp clone of this specific commit around permanently just for your addition, it's that useful.
